docs/source/modifying-messages.rst

```diff
@@ -0,0 +1,390 @@
+`test-invalid-client-hello.py
+<https://github.com/tomato42/tlsfuzzer/blob/master/scripts/test-invalid-client-hello.py>`_
+script.
+
+Modifying records
+=================
+
+The :term:`TLS` protocol specifies four types of encrypted records:
+ones that use stream encryption, ones that use block encryption in
+:term:`MAC` then encrypt mode, ones that use block encryption in encrypt then
+:term:`MAC` mode, and ones that use :term:`AEAD` ciphers.
+Each of them behaves differently on record layer level, thus modifying the
+intermediate ciphertext requires use of different functions.
+
+Fuzzing the MAC
+---------------
+
+Depending on cipher suite and negotiated extensions, to change the
+authentication tag you need to use different functions.
+
+For ciphers that use :term:`HMAC` you can change the authentication tag using
+the :py:func:`~tlsfuzzer.messages.fuzz_mac` function.
+
+.. note::
+
+    :py:func:`~tlsfuzzer.messages.fuzz_mac` works with stream ciphers and
+    block ciphers in :term:`CBC` mode only. It doesn't work for SSLv2
+    connections though.
+
+You use :py:func:`~tlsfuzzer.messages.fuzz_mac` the same way as you use
+:py:func:`~tlsfuzzer.messages.fuzz_message`: pass message to change as the
+first argument and use the other two to specify bytes to either xor or
+substitute.
+
+For example, to invert the first and last bit of :term:`HMAC` in a record with
+a Finished message use the following code:
+
+.. code:: python
+
+    msg_gen = FinishedGenerator()
+    xors = {0: 0x80, -1: 0x01}
+    node = node.add_child(fuzz_mac(msg_gen, xors=xors))
+
+You can find example use of it in
+`test-fuzzed-MAC.py
+<https://github.com/tomato42/tlsfuzzer/blob/master/scripts/test-fuzzed-MAC.py>`_
+script.
+
+Since both :term:`AEAD` cipher suites and :term:`CBC` cipher suites in encrypt
+then :term:`MAC` mode don't encrypt the authentication tag, you can use the
+:py:func:`~tlsfuzzer.messages.fuzz_encrypted_message` function to change it.
+
+As it allows modification of any part of encrypted message, not just the tag,
+you need to know the size of the authentication tag to change the first byte
+of it though.
+
+.. hint::
+
+    :term:`AES-CCM8` uses tags 8 bytes long.
+    :term:`AES-GCM`, Chacha20-Poly1305, :term:`AES-CCM` and MD5-HMAC use tags
+    16 bytes long.
+    SHA1-HMAC uses tags 20 bytes long.
+    SHA256-HMAC uses tags 32 bytes long.
+    SHA384-HMAC uses tags 48 bytes long
+
+For example, to invert the first and last bit of authentication tag in
+a record with a Finished message in an :term:`AES-GCM` connection, use the
+following code:
+
+.. code:: python
+
+    msg_gen = FinishedGenerator()
+    xors = {-17: 0x80, -1: 0x01}
+    node = node.add_child(fuzz_encrypted_message(msg_gen, xors=xors))
+
+You can find example use of this method in
+`test-chacha20.py
+<https://github.com/tomato42/tlsfuzzer/blob/master/scripts/test-chacha20.py>`_
+script.
+
+On the extreme end, tlsfuzzer supports also changing the whole plaintext
```

Comment:
I think with "On the extreme end" you mean to say that "tlsfuzzer can go as far as changing all the plain text".
If that's the case then I would use this quoted text instead.

Reply:
at least wikipedia considers the "plaintext" to be mean specifically the data that is unencrypted while "plain text" to mean human readable unformatted data. Compare [plaintext](https://en.wikipedia.org/wiki/Plaintext) with [plain text](https://en.wikipedia.org/wiki/Plain_text).

otherwise changed as suggested